10up/classifai
# Transcribe audio files using the OpenAI Whisper (speech-to-text) API

### Description of the Change

This PR introduces a new integration with the [OpenAI Whisper](https://platform.openai.com/docs/guides/speech-to-text) (speech-to-text) API. This integration will automatically create a transcript for any supported audio files and store that transcript as the `post_content` for the item (shows in the `Description` field).

### Workflow

A new settings section is added under `Tools > ClassifAI > Language Processing > OpenAI Whisper`. Here there are three options to choose from:

1. Enter your API key
2. Turn the transcription feature on
3. Choose which roles have access

<img width="757" alt="OpenAI Whisper settings" src="https://user-images.githubusercontent.com/916738/236299080-66dcce01-42af-4005-acb7-c92e710608b6.png">

Once configured, whenever a valid audio file is uploaded (must be under 25 MB and the file type has to be one of: mp3, mp4, mpeg, mpga, m4a, wav, or webm) we send that file to the Whisper API and if we get a successful response back, we'll store the transcript as the `post_content` of the attachment item.

For existing audio items, there's a few options to generate transcripts. You can go to the Media Library grid view and click on an audio file. Within the modal that pops up, there will be an option there to (Re-) Transcribe:

<img width="604" alt="Transcribe audio file in media modal" src="https://user-images.githubusercontent.com/916738/236299886-31ed9d38-894f-4dac-84e7-ae58e5a9538b.png">

You can also go to the single media view and there will be a custom metabox with an option to (Re-) Transcribe the item. Check the box and then save the item:

<img width="473" alt="Transcribe audio file in single view" src="https://user-images.githubusercontent.com/916738/236300146-d09e9227-6bff-4191-8fc6-87b6369f4dce.png">

If you prefer using the Media Model list view, there's a `Transcribe audio` bulk edit option as well as an inline `Transcribe` option:

| Bulk transcribe | Inline transcribe |
| --- | --- |
| <img width="379" alt="Bulk transcribe" src="https://user-images.githubusercontent.com/916738/236300423-42580bc0-69e8-436e-bbd9-828af4bcadaa.png"> | <img width="460" alt="Inline transcribe" src="https://user-images.githubusercontent.com/916738/236300591-18f5ea50-a633-4e07-b522-e6e13bcca4d1.png">  |

### How to test the Change

1. Go to `Tools > ClassifAI > Language Processing > OpenAI Whisper` and configure the feature
2. Upload a new audio file that meets the requirements
3. Go to the audio file and ensure the `Description` field has content
4. Test out the other ways to transcribe as explained above, using the media modal, using the single media view and using bulk edit options

### Changelog Entry

> Added - Automatically create transcripts of audio files using the OpenAI Whisper API

### Credits

Props @dkotter

### Checklist:
